File: TypeScript/tester.ts
import * as fs from 'node:fs';
import {solve} from './solver';

fs.readFile('../sudoku.csv', function(err, data) {
    if (err) {
        throw err;
    }
    let lines = data.toString().split('\n');
    let block = 1;
    let start =  Date.now();
    const overall = Date.now();
    for (let i = 0; i < lines.length; i++) {
        let quiz: Number[] = [];
        let answer: Number[] = [];
        if (i == 0) {
            continue;
        }
        let line = lines[i].split(',');
        for (let j = 0; j < line[0].length; j++) {
            quiz.push(Number(line[0][j]));
            answer.push(Number(line[1][j]));
        }
        quiz = solve(quiz);
        if (quiz.toString() !== answer.toString()) {
            console.log("Error on Quiz " + i);
            console.log("Solution found: " + quiz.toString());
            console.log("Solution expected: " + answer.toString());
            return;
        }
        if (i % 50000 == 0 && i >= 50000){
            let ellapsed = Date.now() - start;
            let millis = ellapsed % 1000;
            let seconds = Math.floor(ellapsed / 1000);
            let minutes = Math.floor(seconds / 60);
            console.log("Block " + block + " took " + minutes + " minutes, " + seconds + " seconds, and " + millis + " milliseconds");
            ellapsed = Date.now() - overall;
            millis = ellapsed % 1000;
            seconds = Math.floor(ellapsed / 1000);
            minutes = Math.floor(seconds / 60);
            console.log("Solved " + i + " quizzes in " + minutes + " minutes, " + seconds + " seconds, and " + millis + " milliseconds");
            block++;
            start = Date.now();
        }
    }
});

File: Java/src/Tester.java
import java.io.File;
import java.util.Arrays;
import java.util.Scanner;

public class Tester {
  public static void main(String[] args) {
    try {
      Scanner in = new Scanner(new File("../sudoku.csv"));
      in.nextLine();
      Solver solver = new Solver();
      int block = 1;
      int index = 1;
      long start = System.currentTimeMillis();
      long overall = System.currentTimeMillis();
      while (in.hasNextLine()){
        String line = in.nextLine();
        String[] tokens = line.split(",");
        int[][] quiz = new int[9][9];
        int[][] answer = new int[9][9];
        for (int i = 0; i < 9; i++) {
          for (int j = 0; j < 9; j++) {
            quiz[i][j] = Character.getNumericValue(tokens[0].charAt(i * 9 + j));
            answer[i][j] = Character.getNumericValue(tokens[1].charAt(i * 9 + j));
          }
        }
        int[][] result = solver.Solve(quiz);
        if (!Arrays.deepEquals(result, answer)) {
          System.out.println("Error on Quiz " + index);
          System.out.println("Solution found: ");
          for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
              System.out.print(result[i][j] + " ");
            }
            System.out.println();
          }
          System.out.println("Solution expected: ");
          for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
              System.out.print(answer[i][j] + " ");
            }
            System.out.println();
          }
          return;
        }
        if (index % 50000 == 0 && index >= 50000){
          long ellapsed = System.currentTimeMillis() - start;
          int seconds = (int) (ellapsed / 1000);
          int millis = (int) (ellapsed % 1000);
          System.out.println("Block " + block + " took " + seconds + " seconds and " + millis + " milliseconds");
          ellapsed = System.currentTimeMillis() - overall;
          seconds = (int) (ellapsed / 1000);
          millis = (int) (ellapsed % 1000);
          System.out.println("Solved " + index + " quizzes in " + seconds + " seconds and " + millis + " milliseconds");
          block++;
          start = System.currentTimeMillis();
        }
        index++;
      }
    } catch (Exception e) {
        System.out.println("File not found");
    }
    System.out.println("All quizzes solved");
  }
}
